File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<int> stack;
        unordered_map<char, char> matching {{'(', ')'}, {'{','}'}, {'[',']'}};

        for (char c : s) {
            if (matching.find(c) != matching.end()) {
                stack.push(c);
            }
            else {
                if (stack.empty()) {
                    return false;
                }
                
                char previous_character = stack.top();
                if (matching[previous_character] != c) {
                    return false;
                }
                else {
                    stack.pop();
                }
            }

        }
        return stack.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        matching = {'(':')', '{':'}', '[':']'}

        for c in s:
            if c in matching:
                stack.append(c)
            else:
                if not stack:
                    return False
                previous_char = stack[-1]
                if matching.get(previous_char) != c:
                    return False
                else:
                    stack.pop()
        return len(stack) == 0